[MemProf] Convert Hot contexts to NotCold early

teresajohnson opened the pull request:
While we convert hot contexts to notcold contexts during the cloning
step, their existence was greatly limiting the context trimming
performed when we add the MemProf profile to the IR. To address this,
any hot contexts are converted to notcold contexts immediately after
first checking for unambiguous allocation types, and before checking it
again and before adding metadata while performing context trimming.

Note that hot hints are now disabled by default, however, this avoids
adding unnecessary overhead if they are re-enabled.

teresajohnson committed:
[MemProf] Convert Hot contexts to NotCold early

While we convert hot contexts to notcold contexts during the cloning
step, their existence was greatly limiting the context trimming
performed when we add the MemProf profile to the IR. To address this,
any hot contexts are converted to notcold contexts immediately after
first checking for unambiguous allocation types, and before checking it
again and before adding metadata while performing context trimming.

Note that hot hints are now disabled by default, however, this avoids
adding unnecessary overhead if they are re-enabled.

teresajohnson commented on llvm/test/Transforms/PGOProfile/memprof_loop_unroll.ll:
```diff
@@ -10,7 +10,9 @@
 
 ; RUN: llvm-profdata merge %S/Inputs/memprof_loop_unroll.memprofraw --profiled-binary %S/Inputs/memprof_loop_unroll.exe -o %t.memprofdata
-; RUN: opt < %s -passes='memprof-use<profile-filename=%t.memprofdata>' -S -memprof-report-hinted-sizes 2>&1 | FileCheck %s
+;; Set the minimum lifetime threshold to 0 to ensure that one context is
```
This change is necessary because previously one context was NotCold and the other was Hot, and now they were both NotCold, changing the test behavior.

snehasish commented on llvm/lib/Analysis/MemoryProfileInfo.cpp:
```diff
@@ -222,6 +228,15 @@ void CallStackTrie::collectContextSizeInfo(
 }
 
+void CallStackTrie::convertHotToNotCold(CallStackTrieNode *Node) {
+  if (Node->AllocTypes & static_cast<uint8_t>(AllocationType::Hot)) {
+    Node->AllocTypes &= ~static_cast<uint8_t>(AllocationType::Hot);
+    Node->AllocTypes |= static_cast<uint8_t>(AllocationType::NotCold);
```
Can we add some helpers to Node so that we don't have to sprinkle static_cast and bit masking operations in different places?

teresajohnson replied:
will do

teresajohnson replied:
done

snehasish commented on llvm/lib/Analysis/MemoryProfileInfo.cpp:
```diff
@@ -301,6 +316,22 @@ bool CallStackTrie::buildAndAttachMIBMetadata(CallBase *CI) {
+  // allocation have a single NotCold allocation type), greatly reducing
+  // overheads in bitcode, cloning memory and cloning time.
+  if (Alloc->AllocTypes & static_cast<uint8_t>(AllocationType::Hot)) {
+    convertHotToNotCold(Alloc);
```
I'm not clear on why we need this. If we set `MemProfUseHotHints` to false, then no alloctypes should have hot type and we'll never enter this condition?

teresajohnson replied:
By default hot hint generation is off, however, since it is still optionally supported this avoids a big overhead increase if that is re-enabled for experimentation etc. If we keep that support here, even if optional, I want to ensure things keep working the same.

snehasish replied:
I see, so this change really does 2 things -- optimize when hot hints are used and allow disabling them altogether. The description now makes more sense to me. Could you commit this as two separate changes -- adding the flag and the optimization (conversion to notcold)? For the former it would be good to have a small test for the flag usage.

teresajohnson replied:
ok I can do that. There is a unittest that tests with the flag, but I'll update llvm/test/Transforms/PGOProfile/memprof.ll to test with this as well

teresajohnson replied:
See PR124338. I'll update this one to be just about the conversion once that goes in.

teresajohnson replied:
Updated, ptal

teresajohnson replied:
> For the former it would be good to have a small test for the flag usage.

I had to remove the test I added there as it checks metadata, and I had forgotten that with this change that would no longer happen. I will submit a follow on test that checks the option, using an unambiguously hot allocation.

snehasish reviewed (approved):
lgtm.

teresajohnson committed:
Remove new testing for hot hints added with PR124338 as the conversion
here defeats that. I will add a separate hot hint test for the option as
a follow on.
